### ISSUE TITLE
How to parse templates using front end script.

### ISSUE THREAD

**Issue by avstudios** [dart-bridge/framework#12]
I've checked the templates documentation and you wrote that Handlebars could be used to parse templates. Does it mean there is not build in mechanism to parse templates by the front end scripts at the moment?

**Comment by emilniklas** [comment]
I'm afraid the documentation is kind of misleading at the moment. The proposed workflow (which is in progress) looks like this:

``` dart
// Server script
tether.listen('template', (_) {
  return template('index', withData: {'key': 'value'});
});
```

``` html
<!-- index.html -->
<h1>$key</h1>
```

``` dart
// Client script
Template template = await tether.send('template');
print(template.handlebars); // "<h1>{{ key }}</h1>"
print(template.source); // "<h1>$key</h1>"
print(template.data); // {'key': 'value'}
```

Precompilers should work too:

``` jade
// index.jade
h1= key
```

The above should result in an identical `Template` object on the client.

**Comment by emilniklas** [comment]
As of `1.0.0.alpha1+7`, this is now implemented, though not with complete set of features. Also, the getter is not called `handlebars`. It's `asHandlebars`. And `source` is superfluous, so it was not implemented.

**Comment by marcinsydor** [comment]
Thanks Emil. I've just tested it and I've got the same outputs you. But I'm still not closer to what I want to archive. I would like client script to update e.g. div tag in my main index.html. 
Based on your example I'm getting html as handlebars format. Unfortunately I'm not sure how to use it.
Is there a way that bridge will update my div tag to `<div><h1>value</h1></div>` html?

**Comment by emilniklas** [comment]
I'm afraid Bridge doesn't do that for you. You will have to choose a front end framework, or just code it yourself. There is a small _handlebars_ library that can parse handlebars script, so then you could just update the DOM with that. A tutorial is coming up on how to use Bridge with Polymer as well

**Comment by marcinsydor** [comment]
I see. I've hoped that bridge can do this. I could use angular or polymer for making apps. But for regular websites not sure if that is good idea. 
So far I've made few small websites and used pure dart html api to update the DOM. But I would prefer use some template lib / framework to write html parts. Handlebars might be the solution but can't find a dart version of that script. Where I can get it from?

**Comment by emilniklas** [comment]
https://pub.dartlang.org/packages/mustache

I've also done some work on porting http://vuejs.org, which is a superb lightweight alternative to angular and polymer. When I get the time, I'll keep improving that port (https://pub.dartlang.org/packages/vue) and make sure it works well with Bridge.

Stay tuned :)

**Comment by marcinsydor** [comment]
Thanks for the links. I will keep eye on vue dart version and will try mustache later.
At the moment I'm trying build website using your framework. For now i'm going keep all assets and json data files on separate server. I'm wondering is there a way for dart server scripts to get json files. So I can pass that data to templates in bridge. I can easy use HttpRequest.getString on client side but not sure if there any method to do it on the server side?

**Comment by emilniklas** [comment]
Check out `dart:io.HttpClient` here: https://api.dartlang.org/apidocs/channels/stable/dartdoc-viewer/dart:io.HttpClient

I'm going to release version `1.0.0-alpha2+1` in a second. Please update and report any bugs you might encounter. It's a great help :+1:

Also, I wouldn't recommend relying on Bridge for a production website just yet. Remember – it's an alpha :)

**Comment by marcinsydor** [comment]
Sorry I did not have a chance to check it earlier. Thanks for the link to HttpClient. It works well.
I've updated my project to latest version. I started testing templates and I'm having and error when I make changes to template html file and I'm reloading project.

```
in ShutdownIsolate: 'file:///Users/user/Repos/lab/bridge/.templates.dart': error: line 115 pos 45: ')' expected
  <style>${(await $include('example-styles')}</style>
```

If I remove the line with style all works fine.

I'm doing this website in my free time so it will take me a while to finish it. I hope by that time bridge will be in beta version ;)

**Comment by emilniklas** [comment]
It's great that you are working with a project. Your feedback is golden!

I'm unable to reproduce the problem..

When does this happen? After `new_bridge`? First `dart bridge`? Or after you change the templates? You're not changing `.templates.dart`, right?

**Comment by marcinsydor** [comment]
It's happend after `dart bridge`. I think It stopped working when I used `@for` directive.

```
          @for (model in models)
          <div class="item">
              <h3>${model['model_id']}</h3>
          </div>
          @end for
```

If I remove style line the models are parsed well. Otherwise I have that error. And event I remove the for block then error not disappearing. I have to reset .templates.dart in repo to make it work again.

**Comment by marcinsydor** [comment]
I don't know if that's a bug but if I comment something in template e.g. `<!--<p>$key</p>-->` the line is still executed. I prefer compiler to ignore that line / block. 
Could you also give me quick example how to add template part to Bridge template?

**Comment by emilniklas** [comment]
Okay, there was a regular expression thing. The bug is fixed with `1.0.0-alpha2+2`, but the `Template.asHandlebars` method (which was the source of the error), is still pretty poor.

Feel free to open a new issue if you encounter anymore bugs! :smile:

**Comment by emilniklas** [comment]
**NOTE:** The `asHandlebars` method on `Template` is removed with `1.0.0-alpha3` because it was kind of a hassle. Instead, the recommended approach if you want to send unparsed handlebars to the client, is to create an `.html` template with handlebars, and then send that template through.
